tests/bosonic_files/test_backend_bosoniccircuit.py

```diff
@@ -0,0 +1,406 @@
+        example.reset(2, 1)
+        # Send two coherent states through beamspliter
+        example.displace(r, 0, 0)
+        example.displace(r, 0, 1)
+        example.beamsplitter(phi, phi, 0, 1)
+        alpha = r * (np.cos(phi) - np.exp(-1j * phi) * np.sin(phi))
+        beta = r * (np.cos(phi) + np.exp(1j * phi) * np.sin(phi))
+        assert np.isclose(example.fidelity_coherent(np.array([alpha, beta])), 1)
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    @pytest.mark.parametrize("num_shots", NUM_SHOTS_VALS)
+    def test_generaldyne(self, num_weights, r, num_shots):
+        r"""Checks generaldyne."""
+        # Create two vacuums with multiple weights
+        example = circuit.BosonicModes()
+        example.reset(2, num_weights)
+        # Create squeezed covariance matrix
+        covmat = np.diag(np.array([np.exp(-r), np.exp(r)]))
+        # Send two squeezed states through beamspliter
+        example.squeeze(r, 0, 0)
+        example.squeeze(-r, 0, 1)
+        example.beamsplitter(np.pi / 4, 0, 0, 1)
+        # Perform generaldyne on mode 0
+        vals = example.measure_dyne(covmat, [0], shots=num_shots)
+
+        # Check output has right shape
+        assert vals.shape == (num_shots, 2)
+        # Check measured mode was set to vacuum
+        assert np.isclose(example.fidelity_vacuum(modes=[0]), 1)
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    @pytest.mark.parametrize("phi", PHI_VALS)
+    def test_homodyne(self, num_weights, r, phi):
+        r"""Checks homodyne."""
+        # Create coherent state with multiple weights
+        example = circuit.BosonicModes()
+        example.reset(1, num_weights)
+        # Make complex weights
+
+        example.displace(r, phi, 0)
+        # Perform homodyne on mode 0
+        vals = example.homodyne(0, shots=500)
+        mean = vals[:, 0].mean()
+        std = vals[:, 0].std()
+        assert np.isclose(mean, example.hbar * r * np.cos(phi), rtol=0, atol=10 / np.sqrt(500))
+        assert np.isclose(std, 1, rtol=0, atol=10 / np.sqrt(500))
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    @pytest.mark.parametrize("phi", PHI_VALS)
+    def test_heterodyne(self, num_weights, r, phi):
+        r"""Checks heterodyne."""
+        # Create coherent state with multiple weights
+        example = circuit.BosonicModes()
+        example.reset(1, num_weights)
+        example.displace(r, phi, 0)
+        # Perform heterodyne on mode 0
+        vals = example.heterodyne(0, shots=500)
+        mean = vals[:, 0].mean()
+        assert np.isclose(mean, example.hbar * r * np.cos(phi), rtol=0, atol=10 / np.sqrt(500))
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_post_select_homodyne(self, num_weights, r):
```

Comment:
Could it be useful to also check `post_select_generaldyne` for a non-homodyne/non-heterodyne measurement?

Reply:
I think that would just be another covariance matrix. With homodyne and heterodyne we're already checking two very different cases of covariance matrices so I don't see what else that would add.